lib/pos_snapshot_entries.go

```diff
@@ -0,0 +1,708 @@
+			err,
+			"DBPutSnapshotValidatorEntryWithTxn: problem putting ValidatorEntry in the SnapshotValidatorByPKID index: ",
+		)
+	}
+
+	// Put the ValidatorPKID in the SnapshotValidatorByStake index.
+	key = DBKeyForSnapshotValidatorByStake(validatorEntry, snapshotAtEpochNumber)
+	if err := DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, validatorEntry.ValidatorPKID)); err != nil {
+		return errors.Wrapf(
+			err,
+			"DBPutSnapshotValidatorEntryWithTxn: problem putting ValidatorPKID in the SnapshotValidatorByStake index: ",
+		)
+	}
+
+	return nil
+}
+
+func DBEnumerateAllCurrentValidators(handle *badger.DB, pkidsToSkip []*PKID) ([]*ValidatorEntry, error) {
+	// Convert []*PKIDs of validators to skip to a Set[string] of db keys to skip.
+	skipKeys := NewSet([]string{})
+	for _, pkid := range pkidsToSkip {
+		skipKeys.Add(string(DBKeyForValidatorByPKID(&ValidatorEntry{ValidatorPKID: pkid})))
+	}
+	// Retrieve all non-skipped validators.
+	_, valsFound, err := EnumerateKeysForPrefixWithLimitOffsetOrder(
+		handle, Prefixes.PrefixValidatorByPKID, 0, nil, false, skipKeys,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DBEnumerateValidators: problem retrieving ValidatorEntries")
+	}
+	// Convert ValidatorEntryBytes to ValidatorEntries.
+	var validatorEntries []*ValidatorEntry
+	for _, validatorEntryBytes := range valsFound {
+		validatorEntry := &ValidatorEntry{}
+		rr := bytes.NewReader(validatorEntryBytes)
+		if exist, err := DecodeFromBytes(validatorEntry, rr); !exist || err != nil {
+			return nil, errors.Wrapf(err, "DBEnumerateValidators: problem decoding ValidatorEntry")
+		}
+		validatorEntries = append(validatorEntries, validatorEntry)
+	}
+	return validatorEntries, nil
+}
+
+//
+// SnapshotGlobalActiveStakeAmountNanos
+//
+
+func (bav *UtxoView) GetSnapshotGlobalActiveStakeAmountNanos(snapshotAtEpochNumber uint64) (*uint256.Int, error) {
+	// Check the UtxoView first.
+	if globalActiveStakeAmountNanos, exists := bav.SnapshotGlobalActiveStakeAmountNanos[snapshotAtEpochNumber]; exists {
+		return globalActiveStakeAmountNanos.Clone(), nil
+	}
+	// If we don't have it in the UtxoView, check the db.
+	globalActiveStakeAmountNanos, err := DBGetSnapshotGlobalActiveStakeAmountNanos(bav.Handle, bav.Snapshot, snapshotAtEpochNumber)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"GetSnapshotGlobalActiveStakeAmountNanos: problem retrieving SnapshotGlobalActiveStakeAmountNanos from db: ",
+		)
+	}
+	if globalActiveStakeAmountNanos != nil {
+		// Cache the result in the UtxoView.
+		bav._setSnapshotGlobalActiveStakeAmountNanos(globalActiveStakeAmountNanos, snapshotAtEpochNumber)
+	}
+	return globalActiveStakeAmountNanos, nil
+}
+
+func (bav *UtxoView) _setSnapshotGlobalActiveStakeAmountNanos(globalActiveStakeAmountNanos *uint256.Int, snapshotAtEpochNumber uint64) {
+	if globalActiveStakeAmountNanos == nil {
+		glog.Errorf("_setSnapshotGlobalActiveStakeAmountNanos: called with nil entry, this should never happen")
```

Comment:
```suggestion
		glog.Errorf("_setSnapshotGlobalActiveStakeAmountNanos: called with nil entry, this should never happen")
		return
```
return after this error?